### PR TITLE
otp-nas

### PR DESCRIPTION
**Pull request by ramprasadm-CBT**
-> Fixed error caused by alignment issues in delete slot, slot list, update slot[ no ticket].

-> In register page and profile update page, in terms of OTP,[ticket number #44, Bug 2]

- Maximum number of invalid otp attempt is set as 4 times. after 4 times of invalid attempt the pop up will close automatically.
- Added resent otp count and on trying resend otp 3 times , in 4th attempt, toast error stating “You have reached the maximum number of resend attempts.“ will arise, the form will close automatically.
- cleared otp form input on resend otp, cleared otp on clicking submit again.

-> Added Next available slot to Booking info screen, removed Next available slot details in club list screen, Modified the Next available slot in Featured clubs to display first game's first court's available slot instead of random before.[Ticket number #37
Screenshot (45)
Screenshot (46)
Screenshot (47)
Screenshot (48)
Screenshot (49)
Screenshot (50)
Screenshot (51)
Screenshot (52)
Screenshot (53)
Screenshot (54)
Screenshot (55)
Screenshot (56)
Screenshot (57)
Screenshot (58)
]

### CODE REVIEW

File: backend/frontend/src/screens/BookingInfoScreen.js [hunk 1]
```diff
@@ -339,6 +346,15 @@ function BookingInfoScreen() {
...
   }, [isLogin, userLoginSuccess]);
 
+  
+  useEffect(() => {
+    const theCourt = courts?.find((court) => court.name === courtName);
+    const courtId = theCourt?.id;
+    if (slots?.length !== 0 && additionalSlots?.length !== 0 && courtId) {
+      dispatch(getNearestSlot(courtId, date));
+    }
+  }, [dispatch, courtName, slots, additionalSlots, courts, date]);
```
Comment (TharunCBT) [inline backend/frontend/src/screens/BookingInfoScreen.js]: NIT : Dependencies need to be in Alpabetical order.

File: backend/frontend/src/screens/UpdateProfileScreen.js [hunk 2]
```diff
@@ -107,16 +118,25 @@ function UpdateprofileScreen() {
...
+      } else {
+        toast.error("Incorrect OTP");
+      }
+      setSubmit(false);
     }
     dispatch({
       type: USER_UPDATE_PROFILE_RESET,
     });
-  }, [dispatch, navigate, submit, userUpdateError, userUpdateSuccess]);
+  }, [dispatch, navigate, submit, userUpdateError, userUpdateSuccess, incorrectAttempts]);
```
Comment (TharunCBT) [inline backend/frontend/src/screens/UpdateProfileScreen.js:144]: Dependencies need to be in Alpabetical order.

File: backend/booking/views.py
Comment (TharunCBT) [inline backend/booking/views.py:1]: Same slot functionality here is in serializers.py check that and if possible add a common functionality for it.

**Review by TharunCBT — COMMENTED** [review]
So nit changes and look at that serializer and views same functionality.

File: backend/frontend/src/screens/BookingInfoScreen.js (hunk 1 above)
Comment (ramprasadm-CBT) [reply backend/frontend/src/screens/BookingInfoScreen.js]: done

File: backend/frontend/src/screens/UpdateProfileScreen.js (hunk 2 above)
Comment (ramprasadm-CBT) [reply backend/frontend/src/screens/UpdateProfileScreen.js:144]: done

File: backend/booking/views.py
Comment (ramprasadm-CBT) [reply backend/booking/views.py:1]: Looks about the same, but serializer works with current date and first game and first court, where as view works with selected date dispatched from Booking Info Page, selected game and selected court.

**Review by TharunCBT — APPROVED** [review]
LGTM

File: backend/booking/serializers.py [hunk 3]
```diff
@@ -159,6 +159,7 @@ class Meta:
         model = Organization
         fields = '__all__'
 
+from django.db.models import Q
```
Comment (murucbt) [inline backend/booking/serializers.py:163]: Why are we importing in middle of the file?

File: backend/booking/serializers.py [hunk 4]
```diff
@@ -191,41 +192,108 @@ def get_reviews(self, obj):
...
+
+        # Find nearest slot from Slot table
+        for i in range(0, 7):  # Search for up to one week
+            target_date = now + datetime.timedelta(days=i)
+            target_weekday = target_date.strftime('%A')
+
+            slots = Slot.objects.filter(
+                court=court,
+                days=target_weekday,
+                start_time__gte=(current_time if i == 0 and selected_date == now.date() else datetime.datetime.min.time())
```
Comment (murucbt) [inline backend/booking/serializers.py]: Don't we have is_active flag here for Slot? If yes, then both Slot & AdditionalSlot forloop can be combined in a single function (may be util) and use that here!

File: backend/booking/views.py [hunk 5]
```diff
@@ -572,3 +572,113 @@ def getHighRatedClubs(request):
...
+        )
+
+        for slot in additional_slots:
+            slot_info = {
+                'date': slot.date,
+                'start_time': slot.start_time,
+                'end_time': slot.end_time,
+                'source': 'additional_slot'
+            }
+            nearest_slots.append(slot_info)
```
Comment (murucbt) [inline backend/booking/views.py]: Seem this block of code is repeating in multiple places, think of reducing duplicates.

File: backend/frontend/src/screens/BookingInfoScreen.js [hunk 6]
```diff
@@ -500,11 +516,15 @@ function BookingInfoScreen() {
...
                         }))}
                       />
-                    ) : (
-                      <Alert severity="error">
-                      {t("noSlotsAvailable", {courtName})}                       
-                      </Alert>
-                    )}
+                    ) : nearestSlot ? (
+                        <Alert severity="info">
+                          Currently No slots available in {courtName}, Next slot available on {nearestSlot?.days ? nearestSlot.days : formatDate(nearestSlot.date)} at {nearestSlot?.start_time?.slice(0, 5)}
```
Comment (murucbt) [inline backend/frontend/src/screens/BookingInfoScreen.js]: This has to be translated.

**Review by murucbt — CHANGES_REQUESTED** [review]

File: backend/frontend/src/screens/BookingInfoScreen.js (hunk 6 above)
Comment (ramprasadm-CBT) [reply backend/frontend/src/screens/BookingInfoScreen.js]: translated and ready.

File: backend/booking/serializers.py (hunk 4 above)
Comment (ramprasadm-CBT) [reply backend/booking/serializers.py]: slot follow days and additional slot follow date, slot has is_booked flag yes, Ok will create and use util function.

File: backend/booking/serializers.py (hunk 3 above)
Comment (ramprasadm-CBT) [reply backend/booking/serializers.py:163]: will move it now. will make sure this mistake will not be repeated again.

**Review by murucbt — APPROVED** [review]

**Pull request merged**